File: August2022/Java/PalindromeLinkedList.java

// @saorav21994
// TC : O(n)
// SC : O(1)

// in-place

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        
        ListNode slow = head;
        ListNode fast = head;
        ListNode prev = null;
        
        if (head == null || head.next == null)
            return true;
        
        
        // find mid point
        while (fast != null && fast.next != null) {
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        
        prev.next = null;
        
        ListNode curr = slow;
        prev = null;
        
        // reverse the 2nd half
        while (curr != null) {
            ListNode tmp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = tmp;
        }
        
        slow = prev;
        
        // compare for palindrome
        
        while (head != null && slow != null) {
            if (head.val != slow.val)
                return false;
            head = head.next;
            slow = slow.next;
        }
        
        return true;
        
    }
}

File: August2022/C++/PalindromeLinkedList.cpp
class Solution {
private:
    ListNode* reverse(ListNode* root) {
        ListNode* prev = NULL;
        ListNode* curr = root;
        
        while (curr != NULL) {
            ListNode* next = curr->next;
            curr->next = prev;            
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
    
public:
    // TC: O(n)
    // SC: O(1)
    bool isPalindrome(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;
        
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;            
        }
        
        if (fast != NULL) { // it means there are odd nodes in the linked-list
            slow = slow->next;
        }
        
        slow = reverse(slow);
        fast = head;
        
        while (slow != NULL) {
            if (fast->val != slow->val) {
                return false;
            }
            fast = fast->next;
            slow = slow->next;
        }
                
        return true;
    }
};
